refactor: Do not discard `try_lock()` return value

hebasto opened the pull request:
Microsoft's C++ Standard Library uses the `[[nodiscard]]` attribute for `try_lock()`.
See: https://github.com/microsoft/STL/blob/main/stl/inc/mutex

This change allows to drop the current suppression for the warning C4838 and helps to prevent the upcoming warning C4858.
See: https://github.com/microsoft/STL/commit/539c26c923b38cd0b5eba2bb11de4bea9d5c6e43

Fixes bitcoin/bitcoin#26017.

Split from bitcoin/bitcoin#25819.

hebasto committed:
refactor: Do not discard `try_lock()` return value

Microsoft's C++ Standard Library uses the `[[nodiscard]]` attribute for
`try_lock()`.
See: https://github.com/microsoft/STL/blob/main/stl/inc/mutex

This change allows to drop the current suppression for the warning C4838
and helps to prevent the upcoming warning C4858.
See: https://github.com/microsoft/STL/commit/539c26c923b38cd0b5eba2bb11de4bea9d5c6e43

hebasto commented:
Friendly ping @ajtowns @ryanofsky @vasild

ajtowns commented:
Concept ACK.

This seems to have been introduced in 712fd182b72b0f5a1bcf843f171c29ec0a49b50f (April 2012) by @sipa -- the old code for `TryEnter` did `result = mutex.try_lock(); if (!result) pop_lock(); return result;` the new code discarded the result for `try_lock()` and instead then checked `!lock.owns()`.

vasild reviewed (approved):
ACK bff4e068b69edd40a00466156f860bde2df29268

The old and new code should have identical behavior. The new code is shorter and silences a warning.

https://en.cppreference.com/w/cpp/thread/unique_lock/try_lock
https://en.cppreference.com/w/cpp/thread/unique_lock/owns_lock

The other call to `Base::owns_lock()` can also be avoided:

```cpp
        EnterCritical(pszName, pszFile, nLine, Base::mutex(), true);
        if (!Base::try_lock()) {
            LeaveCritical();
            return false;
        }     
        return true;
```

sipa reviewed (commented):
ACK bff4e068b69edd40a00466156f860bde2df29268

I'd suggest incorporating @vasild's suggestion above.

hebasto committed:
refactor: Drop `owns_lock()` call

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

hebasto commented:
Updated bff4e068b69edd40a00466156f860bde2df29268 -> 30cc1c6609ad7868f73e88afe0b0233d395ec08c ([pr26189.01](https://github.com/hebasto/bitcoin/commits/pr26189.01) -> [pr26189.02](https://github.com/hebasto/bitcoin/commits/pr26189.02), [diff](https://github.com/hebasto/bitcoin/compare/pr26189.01..pr26189.02)):

- added a commit which implements @vasild's [suggestion](https://github.com/bitcoin/bitcoin/pull/26189#pullrequestreview-1123259931)

vasild reviewed (approved):
ACK 30cc1c6609ad7868f73e88afe0b0233d395ec08c
